### ISSUE TITLE
AOT inductor can't handle dynamic shape 2*s1

### ISSUE THREAD

**Issue by henrylhtsang** [pytorch/pytorch#137748]
### 🐛 Describe the bug

Internal post: https://fb.workplace.com/groups/1028545332188949/posts/1063267995383349/

Hi. I was looking at a case where it is beneficial to tell torch.export a dim is a multiple of 2. So I did dynamic_shapes={"x": {1: 2 * bs}}. Export runs fine.

However, when I tried to aot compile, I ran into C++ compilation error, which says s0 is not defined. The problem is, while s1 is a [sympy.Symbol](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/wrapper.py#L1052), 2*s1 is a <class 'sympy.core.mul.Mul'> instead.

repro:
```
import torch 
class TestModel(torch.nn.Module):
    def forward(self, x):
        return x.shape[1]

from torch.export import Dim

model = TestModel().cuda()
x = torch.rand(2, 6, 6).cuda()
args = (x,)

bs = Dim("bs", max=128)
ep = torch.export.export(
    model, args, dynamic_shapes={"x": {1: 2 * bs}}, strict=False
)

aot_model = torch._inductor.aot_compile(ep.module(), args)
```

codegen: (Note that s1 is not defined)
```
auto inputs = steal_from_raw_handles_to_raii_handles(input_handles, 1);
auto arg0_1 = std::move(inputs[0]);
inputs.clear();
auto& kernels = static_cast<AOTInductorModelKernels&>(*this->kernels_.get());
RAIIAtenTensorHandle scalar_to_tensor_0 = scalar_to_tensor_handle(2L*s1);
output_handles[0] = scalar_to_tensor_0.release();
```

C++ compilation error:
```
 error: use of undeclared identifier 's1'
  593 |     RAIIAtenTensorHandle scalar_to_tensor_0 = scalar_to_tensor_handle(2L*s1);
      |                                                                          ^
1 error generated.
```

### Versions

trunk

cc @ezyang @chauhang @penguinwu @bobrenjc93 @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4 @desertfire @chenyang78

**Comment by ColinPeppler** [comment]
From the looks of it, `s1` isn't being codegen-ed hence the error?

**Comment by henrylhtsang** [comment]
> From the looks of it, `s1` isn't being codegen-ed hence the error?

Yes

> EDIT: I guess I see why, is there no input whose size is strictly s1?

what?

**Comment by ezyang** [comment]
Oh yeah, you can't do this. Specifically, export shouldn't let you do this. A key invariant in our codegen is that every dynamic shape must exist as an actual argument to the program; I think export relaxes this somewhat but it still doesn't let you do what you did here.

The easiest fix is to explicitly also pass in bs as an argument, so that it gets bound.  Not really sure what other constraints you're operating under though.

**Comment by henrylhtsang** [comment]
> Oh yeah, you can't do this. Specifically, export shouldn't let you do this. A key invariant in our codegen is that every dynamic shape must exist as an actual argument to the program; I think export relaxes this somewhat but it still doesn't let you do what you did here.
> 
> The easiest fix is to explicitly also pass in bs as an argument, so that it gets bound. Not really sure what other constraints you're operating under though.

The reason I want to do it is specify my dynamic shape as something that is divisible by 8. If I don't do that in export, I will get an error from export/dynamo:
```
  - Not all values of pic_d1 = L['args'][0][0].size()[2] in the specified range 56 <= pic_d1 <= 96 satisfy the generated guard Ne(Mod(1, ((L['args'][0][0].size()[2] - 1)//2) + 1), 0).
  - Not all values of pic_d1 = L['args'][0][0].size()[2] in the specified range 56 <= pic_d1 <= 96 satisfy the generated guard Ne(Mod(1, 8*(((L['args'][0][0].size()[2] - 1)//8)) + 8), 0).
Suggested fixes:
  _pic_d1 = Dim('_pic_d1', min=7, max=12)
  _pic_d2 = Dim('_pic_d2', min=7, max=12)
  pic_d1 = 8*_pic_d1
  pic_d2 = 8*_pic_d2
```

**Comment by ezyang** [comment]
Yeah, the suggested fix is wrong (@avikchaudhuri). Honestly we should just get export to shut up about this, there's nothing wrong with an implicit divisibility constraint......

**Comment by henrylhtsang** [comment]
This is bothering me a lot. I was trying to run another example, but export keeps throwing an error because of divisibility problem. Repro

```
import torch
class TestModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x, y=None):
        return x + y.reshape(x.shape)

model = TestModule().cuda()
x = torch.ones(4, 4).cuda()
y = torch.rand(2, 8).cuda()

model(x, y)
from torch.export import Dim

d0 = Dim("d0")
d1 = Dim("d1")

ep = torch.export.export(
    model, (x, y), dynamic_shapes={"x": None, "y": {0: d0, 1: d1}}, strict=False
)
```

Error:
```
UserError:Constraints violated (d0)! For more information, run with TORCH_LOGS="+dynamic".
  - Not all values of d0 = L['args'][0][1].size()[0] in the specified range satisfy the generated guard Ne(L['args'][0][1].size()[0], 4).
  - Not all values of d0 = L['args'][0][1].size()[0] in the specified range satisfy the generated guard Ne(Mod(L['args'][0][1].size()[0], 4), 0).
```

**Comment by pianpwk** [comment]
@henrylhtsang could you switch to using `Dim.AUTO` or `Dim.DYNAMIC` markers for those dims instead? They should have a much better time getting past these complex guards. For example:
```
ep = export(
    model, (x, y), dynamic_shapes={"x": None, "y": {0: Dim.AUTO, 1: Dim.AUTO}}, strict=False
)
print(ep)
```

This gets me:
```
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, x: "f32[4, 4]", y: "f32[s0, s1]"):
             # 
            sym_size_int_3: "Sym(s0)" = torch.ops.aten.sym_size.int(y, 0)
            sym_size_int_4: "Sym(s1)" = torch.ops.aten.sym_size.int(y, 1)
            
             # File: /data/users/pianpwk/pytorch/test/export/test_export.py:2798 in forward, code: return x + y.reshape(x.shape)
            view: "f32[4, ((s0*s1)//4)]" = torch.ops.aten.view.default(y, [4, 4]);  y = None
            
             # 
            mul_4: "Sym(s0*s1)" = sym_size_int_3 * sym_size_int_4;  sym_size_int_3 = sym_size_int_4 = None
            eq_3: "Sym(Eq(s0*s1, 16))" = mul_4 == 16
            _assert_scalar_default = torch.ops.aten._assert_scalar.default(eq_3, "Runtime assertion failed for expression Eq(s0*s1, 16) on node 'eq_3'");  eq_3 = _assert_scalar_default = None
            floordiv: "Sym(((s0*s1)//4))" = mul_4 // 4;  mul_4 = None
            eq_4: "Sym(Eq(((s0*s1)//4), 4))" = floordiv == 4;  floordiv = None
            _assert_scalar_default_1 = torch.ops.aten._assert_scalar.default(eq_4, "Runtime assertion failed for expression Eq(((s0*s1)//4), 4) on node 'eq_4'");  eq_4 = _assert_scalar_default_1 = None
            
             # File: /data/users/pianpwk/pytorch/test/export/test_export.py:2798 in forward, code: return x + y.reshape(x.shape)
            add_3: "f32[4, 4]" = torch.ops.aten.add.Tensor(x, view);  x = view = None
            return (add_3,)
            
Graph signature: ExportGraphSignature(input_specs=[InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='x'), target=None, persistent=None), InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='y'), target=None, persistent=None)], output_specs=[OutputSpec(kind=<OutputKind.USER_OUTPUT: 1>, arg=TensorArgument(name='add_3'), target=None)])
Range constraints: {s0: VR[2, int_oo], s1: VR[2, int_oo]}
```

**Comment by henrylhtsang** [comment]
@pianpwk thanks. Some questions:
1. what is the difference between Dim.AUTO and Dim.DYNAMIC? Do they specialize if dim is 0 or 1?
2. I am getting a thousand lines of warning logs "Ignored guard Eq(s0*s1, 16) == True, this could result in accuracy problems" coming from "torch/fx/experimental/symbolic_shapes.py". Can reduce the log level?

**Comment by pianpwk** [comment]
@henrylhtsang I'm working on improving the documentation, hopefully this should go into Export's dynamic shapes tutorial soon. `AUTO` doesn't complain if the shape specializes (model tracing infers it to be constant, not dynamic), while `DYNAMIC` complains in this case. Under the hood they lower to `torch._dynamo.maybe_mark_dynamic()` and `mark_dynamic()` respectively, there's some details in the documentation there.

> 2. I am getting a thousand lines of warning logs "Ignored guard Eq(s0*s1, 16) == True

Not sure, let me look into this.. seems like it's regarding runtime asserts being frozen in inductor lowering.

**Comment by bhack** [comment]
> I'm working on improving the documentation, hopefully this should go into Export's dynamic shapes tutorial soon. 

@pianpwk Can you ping me as soon as you have this PR?

**Comment by henrylhtsang** [comment]
@pianpwk Just to recap:
1. I was having a different problem, and I switched to Dim.DYNAMIC, and it is much better.
2. The excessive warning logging problem still is there. repro:
```

class TestModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x, y=None):
        return x + y.reshape(x.shape)

model = TestModule().cuda()
x = torch.ones(4, 4).cuda()
y = torch.rand(2, 8).cuda()

model(x, y)
from torch.export import Dim

_ = torch.export.export(
    model,
    (x, y),
    dynamic_shapes={"x": None, "y": {0: Dim.DYNAMIC, 1: Dim.DYNAMIC}},
    strict=False,
)
```

**Comment by henrylhtsang** [comment]
I will split the excessive warning issue into a separate issue and close this one.
